experiment/simulation/js/fa.js

```diff
@@ -0,0 +1,333 @@
+        registerGate(this.id, this);
+    }
+
+    // Sets values of the inputs and outputs of the full adder
+    setA0(A0) {
+        this.a0 = A0;
+    }
+
+    setB0(B0) {
+        this.b0 = B0;
+    }
+
+    setCin(Cin) {
+        this.Cin = Cin;
+    }
+
+    setSum(Sum) {
+        this.sum = Sum;
+    }
+
+    setCout(Cout) {
+        this.Cout = Cout;
+    }
+
+    // adds input endpoints points to the list of input points
+    addInputPoints(input) {
+        this.inputPoints.push(input);
+    }
+
+    // Adds the output endpoints to the list of output points
+    addOutputPoints(output) {
+        this.outputPoints.push(output);
+    }
+
+    // Generates the output of the full adder
+    generateOutput() {
+        const aXorb = (getOutputFA(this.a0[0], this.a0[1]) && !getOutputFA(this.b0[0], this.b0[1])) || (!getOutputFA(this.a0[0], this.a0[1]) && getOutputFA(this.b0[0], this.b0[1]));
+        this.Cout = (getOutputFA(this.a0[0], this.a0[1]) && getOutputFA(this.b0[0], this.b0[1])) || (getOutputFA(this.Cin[0], this.Cin[1]) && aXorb);
+        this.sum = (getOutputFA(this.Cin[0], this.Cin[1]) && !aXorb) || (!(getOutputFA(this.Cin[0], this.Cin[1])) && aXorb);
+    }
+
+    // Sets the output enpoint of the full adder as connected
+    setConnected(val, pos) {
+        console.log(val, pos);
+        if (pos == "Carry") {
+            this.CoutIsConnected = val;
+        }
+        else if (pos == "Sum") {
+            this.sumIsConnected = val;
+        }
+    }
+
+}
+
+// Add a full adder to the circuit board
+function addFA() {
+    let fA = new FullAdder();
+    fA.registerComponent("working-area");
+}
+
+window.AddFA = addFA;
+
+// Used to extract output from a given gate, if pos isnt empty the gate is a full adder with the position specified
+function getOutputFA(gate, pos) {
+    if (pos == "Carry") {
+        return gate.Cout;
+    }
+    else if (pos == "Sum") {
+        return gate.sum;
+    }
+    // But if the gate is not an FA, but an input bit, then return the value of the input
+    else {
+        return gate.output;
+    }
+}
+
+
+
+// Recursive function that evaluates the output of the full adder
+function getResultFA(fa) {
+// check if fa type is Gate object
+    if (fa.constructor.name == "Gate") {
+        return;
+    }
+
+    if (fa.Cout != null && fa.sum != null) {
+        return;
+    }
+
+    if (getOutputFA(fa.a0[0], fa.a0[1]) == null) {
+        getResultFA(fa.a0[0]);
+    }
+    if (getOutputFA(fa.b0[0], fa.b0[1]) == null) {
+        getResultFA(fa.b0[0]);
+    }
+    if (getOutputFA(fa.Cin[0], fa.Cin[1]) == null) {
+        getResultFA(fa.Cin[0]);
+    }
+
+    fa.generateOutput();
+
+    return;
+}
+
+
+// Checks if the connections are correct
+function checkConnectionsFA() {
+    let flag = 0;
+    for (let faID in fullAdder) {
+        const gate = fullAdder[faID];
+        // For Full Adder objects
+        // Check if all the outputs are connected
+        if (gate.CoutIsConnected == false) {

```

Comment:
I believe you can do all null checks in a single line,
and you can directly use
```!gate.CoutIsConnected``` instead of ```gate.CoutIsConnected==false```